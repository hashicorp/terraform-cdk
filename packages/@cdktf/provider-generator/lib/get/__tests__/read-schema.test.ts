import * as path from "path";
// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import {
  readProviderSchema,
  readModuleSchema,
} from "../generator/provider-schema";
import {
  ConstructsMakerModuleTarget,
  ConstructsMakerProviderTarget,
} from "../constructs-maker";
import {
  TerraformModuleConstraint,
  TerraformProviderConstraint,
  Language,
} from "@cdktf/commons";

expect.addSnapshotSerializer({
  test: (value) => {
    if (typeof value !== "object" || value === null) return false;

    return value["format_version"] !== undefined;
  },
  print: (value: any) => {
    value["format_version"] = "STUBBED VERSION";
    return JSON.stringify(value, null, 2);
  },
});

describe("readSchema", () => {
  it("generates a single provider schema", async () => {
    const provider = new TerraformProviderConstraint("hashicorp/null@3.1.0");
    const target = new ConstructsMakerProviderTarget(
      provider,
      Language.TYPESCRIPT
    );
    const result = await readProviderSchema(target);
    expect(result).toMatchSnapshot();
  }, 120000);

  it("generates a single module schema", async () => {
    const module = new TerraformModuleConstraint(
      "terraform-aws-modules/iam/aws//modules/iam-account@3.12.0"
    );
    const target = new ConstructsMakerModuleTarget(module, Language.TYPESCRIPT);
    const result = await readModuleSchema(target);
    expect(result).toMatchSnapshot();
  }, 120000);

  it("generates a more complex schema", async () => {
    const module = new TerraformModuleConstraint(
      "terraform-aws-modules/eks/aws@7.0.1"
    );
    const target = new ConstructsMakerModuleTarget(module, Language.TYPESCRIPT);
    const result = await readModuleSchema(target);
    expect(result).toMatchSnapshot();
  }, 120000);

  it("generates a local module", async () => {
    const module = new TerraformModuleConstraint({
      name: "local_module",
      fqn: "local_module",
      source: path.resolve(__dirname, "fixtures", "local-module"),
    });
    const target = new ConstructsMakerModuleTarget(module, Language.TYPESCRIPT);
    const result = await readModuleSchema(target);
    expect(result).toMatchSnapshot();
  }, 120000);

  it("generates a local json module", async () => {
    const module = new TerraformModuleConstraint({
      name: "local_module",
      fqn: "local_module",
      source: path.resolve(__dirname, "fixtures", "local-json-module"),
    });
    const target = new ConstructsMakerModuleTarget(module, Language.TYPESCRIPT);
    const result = await readModuleSchema(target);
    expect(result).toMatchSnapshot();
  }, 120000);
});
