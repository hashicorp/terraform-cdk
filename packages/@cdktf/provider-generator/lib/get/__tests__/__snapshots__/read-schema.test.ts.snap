// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readSchema generates a local json module 1`] = `
{
  "local_module": {
    "inputs": [
      {
        "default": "module",
        "description": "Name of the DigitalOcean static site",
        "name": "name",
        "required": false,
        "type": "any",
      },
      {
        "default": "ams",
        "description": "Region to deploy the app to",
        "name": "region",
        "required": false,
        "type": "any",
      },
    ],
    "name": "local_module",
    "outputs": [
      {
        "description": undefined,
        "name": "uri",
      },
    ],
  },
}
`;

exports[`readSchema generates a local module 1`] = `
{
  "local_module": {
    "inputs": [
      {
        "description": "Name of the s3 bucket. Must be unique.",
        "name": "bucket_name",
        "required": true,
        "type": "string",
      },
      {
        "default": {},
        "description": "Tags to set on the bucket.",
        "name": "tags",
        "required": false,
        "type": "map(string)",
      },
    ],
    "name": "local_module",
    "outputs": [
      {
        "description": "ARN of the bucket",
        "name": "arn",
      },
      {
        "description": "Domain name of the bucket",
        "name": "domain",
      },
      {
        "description": "Name (id) of the bucket",
        "name": "name",
      },
    ],
  },
}
`;

exports[`readSchema generates a more complex schema 1`] = `
{
  "terraform-aws-modules_eks_aws": {
    "inputs": [
      {
        "default": true,
        "description": "Whether to attach the module managed cluster autoscaling iam policy to the default worker IAM role. This requires \`manage_worker_autoscaling_policy = true\`",
        "name": "attach_worker_autoscaling_policy",
        "required": false,
        "type": "bool",
      },
      {
        "default": true,
        "description": "Whether to attach the Amazon managed \`AmazonEKS_CNI_Policy\` IAM policy to the default worker IAM role. WARNING: If set \`false\` the permissions must be assigned to the \`aws-node\` DaemonSet pods via another method or nodes will not be able to join the cluster.",
        "name": "attach_worker_cni_policy",
        "required": false,
        "type": "bool",
      },
      {
        "default": true,
        "description": "Whether to create a security group for the cluster or attach the cluster to \`cluster_security_group_id\`.",
        "name": "cluster_create_security_group",
        "required": false,
        "type": "bool",
      },
      {
        "default": "15m",
        "description": "Timeout value when creating the EKS cluster.",
        "name": "cluster_create_timeout",
        "required": false,
        "type": "string",
      },
      {
        "default": "15m",
        "description": "Timeout value when deleting the EKS cluster.",
        "name": "cluster_delete_timeout",
        "required": false,
        "type": "string",
      },
      {
        "default": [],
        "description": "A list of the desired control plane logging to enable. For more information, see Amazon EKS Control Plane Logging documentation (https://docs.aws.amazon.com/eks/latest/userguide/control-plane-logs.html)",
        "name": "cluster_enabled_log_types",
        "required": false,
        "type": "list(string)",
      },
      {
        "default": false,
        "description": "Indicates whether or not the Amazon EKS private API server endpoint is enabled.",
        "name": "cluster_endpoint_private_access",
        "required": false,
        "type": "bool",
      },
      {
        "default": true,
        "description": "Indicates whether or not the Amazon EKS public API server endpoint is enabled.",
        "name": "cluster_endpoint_public_access",
        "required": false,
        "type": "bool",
      },
      {
        "default": "",
        "description": "IAM role name for the cluster. Only applicable if manage_cluster_iam_resources is set to false.",
        "name": "cluster_iam_role_name",
        "required": false,
        "type": "string",
      },
      {
        "default": "",
        "description": "If a KMS Key ARN is set, this key will be used to encrypt the corresponding log group. Please be sure that the KMS Key has an appropriate key policy (https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/encrypt-log-data-kms.html)",
        "name": "cluster_log_kms_key_id",
        "required": false,
        "type": "string",
      },
      {
        "default": 90,
        "description": "Number of days to retain log events. Default retention - 90 days.",
        "name": "cluster_log_retention_in_days",
        "required": false,
        "type": "number",
      },
      {
        "description": "Name of the EKS cluster. Also used as a prefix in names of related resources.",
        "name": "cluster_name",
        "required": true,
        "type": "string",
      },
      {
        "default": "",
        "description": "If provided, the EKS cluster will be attached to this security group. If not given, a security group will be created with necessary ingress/egress to work with the workers",
        "name": "cluster_security_group_id",
        "required": false,
        "type": "string",
      },
      {
        "default": "1.14",
        "description": "Kubernetes version to use for the EKS cluster.",
        "name": "cluster_version",
        "required": false,
        "type": "string",
      },
      {
        "default": "./",
        "description": "Where to save the Kubectl config file (if \`write_kubeconfig = true\`). Assumed to be a directory if the value ends with a forward slash \`/\`.",
        "name": "config_output_path",
        "required": false,
        "type": "string",
      },
      {
        "default": "/",
        "description": "If provided, all IAM roles will be created on this path.",
        "name": "iam_path",
        "required": false,
        "type": "string",
      },
      {
        "default": [],
        "description": "Any additional arguments to pass to the authenticator such as the role to assume. e.g. ["-r", "MyEksRole"].",
        "name": "kubeconfig_aws_authenticator_additional_args",
        "required": false,
        "type": "list(string)",
      },
      {
        "default": "aws-iam-authenticator",
        "description": "Command to use to fetch AWS EKS credentials.",
        "name": "kubeconfig_aws_authenticator_command",
        "required": false,
        "type": "string",
      },
      {
        "default": [],
        "description": "Default arguments passed to the authenticator command. Defaults to [token -i $cluster_name].",
        "name": "kubeconfig_aws_authenticator_command_args",
        "required": false,
        "type": "list(string)",
      },
      {
        "default": {},
        "description": "Environment variables that should be used when executing the authenticator. e.g. { AWS_PROFILE = "eks"}.",
        "name": "kubeconfig_aws_authenticator_env_variables",
        "required": false,
        "type": "map(string)",
      },
      {
        "default": "",
        "description": "Override the default name used for items kubeconfig.",
        "name": "kubeconfig_name",
        "required": false,
        "type": "string",
      },
      {
        "default": [
          "/bin/sh",
          "-c",
        ],
        "description": "Command to run for local-exec resources. Must be a shell-style interpreter. If you are on Windows Git Bash is a good choice.",
        "name": "local_exec_interpreter",
        "required": false,
        "type": "list(string)",
      },
      {
        "default": true,
        "description": "Whether to apply the aws-auth configmap file.",
        "name": "manage_aws_auth",
        "required": false,
        "type": "bool",
      },
      {
        "default": true,
        "description": "Whether to let the module manage cluster IAM resources. If set to false, cluster_iam_role_name must be specified.",
        "name": "manage_cluster_iam_resources",
        "required": false,
        "type": "bool",
      },
      {
        "default": true,
        "description": "Whether to let the module manage the cluster autoscaling iam policy.",
        "name": "manage_worker_autoscaling_policy",
        "required": false,
        "type": "bool",
      },
      {
        "default": true,
        "description": "Whether to let the module manage worker IAM resources. If set to false, iam_instance_profile_name must be specified for workers.",
        "name": "manage_worker_iam_resources",
        "required": false,
        "type": "bool",
      },
      {
        "default": [],
        "description": "Additional AWS account numbers to add to the aws-auth configmap. See examples/basic/variables.tf for example format.",
        "name": "map_accounts",
        "required": false,
        "type": "list(string)",
      },
      {
        "default": [],
        "description": "Additional IAM roles to add to the aws-auth configmap. See examples/basic/variables.tf for example format.",
        "name": "map_roles",
        "required": false,
        "type": "any",
      },
      {
        "default": [],
        "description": "Additional IAM users to add to the aws-auth configmap. See examples/basic/variables.tf for example format.",
        "name": "map_users",
        "required": false,
        "type": "any",
      },
      {
        "default": null,
        "description": "If provided, all IAM roles will be created with this permissions boundary attached.",
        "name": "permissions_boundary",
        "required": false,
        "type": "string",
      },
      {
        "description": "A list of subnets to place the EKS cluster and workers within.",
        "name": "subnets",
        "required": true,
        "type": "list(string)",
      },
      {
        "default": {},
        "description": "A map of tags to add to all resources.",
        "name": "tags",
        "required": false,
        "type": "map(string)",
      },
      {
        "description": "VPC where the cluster and workers will be deployed.",
        "name": "vpc_id",
        "required": true,
        "type": "string",
      },
      {
        "default": [],
        "description": "A list of additional security group ids to attach to worker instances",
        "name": "worker_additional_security_group_ids",
        "required": false,
        "type": "list(string)",
      },
      {
        "default": "",
        "description": "Name filter for AWS EKS worker AMI. If not provided, the latest official AMI for the specified 'cluster_version' is used.",
        "name": "worker_ami_name_filter",
        "required": false,
        "type": "string",
      },
      {
        "default": "",
        "description": "Name filter for AWS EKS Windows worker AMI. If not provided, the latest official AMI for the specified 'cluster_version' is used.",
        "name": "worker_ami_name_filter_windows",
        "required": false,
        "type": "string",
      },
      {
        "default": "602401143452",
        "description": "The ID of the owner for the AMI to use for the AWS EKS workers. Valid values are an AWS account ID, 'self' (the current account), or an AWS owner alias (e.g. 'amazon', 'aws-marketplace', 'microsoft').",
        "name": "worker_ami_owner_id",
        "required": false,
        "type": "string",
      },
      {
        "default": "801119661308",
        "description": "The ID of the owner for the AMI to use for the AWS EKS Windows workers. Valid values are an AWS account ID, 'self' (the current account), or an AWS owner alias (e.g. 'amazon', 'aws-marketplace', 'microsoft').",
        "name": "worker_ami_owner_id_windows",
        "required": false,
        "type": "string",
      },
      {
        "default": false,
        "description": "Whether to create initial lifecycle hooks provided in worker groups.",
        "name": "worker_create_initial_lifecycle_hooks",
        "required": false,
        "type": "bool",
      },
      {
        "default": true,
        "description": "Whether to create a security group for the workers or attach the workers to \`worker_security_group_id\`.",
        "name": "worker_create_security_group",
        "required": false,
        "type": "bool",
      },
      {
        "default": [],
        "description": "A list of maps defining worker group configurations to be defined using AWS Launch Configurations. See workers_group_defaults for valid keys.",
        "name": "worker_groups",
        "required": false,
        "type": "any",
      },
      {
        "default": [],
        "description": "A list of maps defining worker group configurations to be defined using AWS Launch Templates. See workers_group_defaults for valid keys.",
        "name": "worker_groups_launch_template",
        "required": false,
        "type": "any",
      },
      {
        "default": "",
        "description": "If provided, all workers will be attached to this security group. If not given, a security group will be created with necessary ingress/egress to work with the EKS cluster.",
        "name": "worker_security_group_id",
        "required": false,
        "type": "string",
      },
      {
        "default": 1025,
        "description": "Minimum port number from which pods will accept communication. Must be changed to a lower value if some pods in your cluster will expose a port lower than 1025 (e.g. 22, 80, or 443).",
        "name": "worker_sg_ingress_from_port",
        "required": false,
        "type": "number",
      },
      {
        "default": [],
        "description": "Additional policies to be added to workers",
        "name": "workers_additional_policies",
        "required": false,
        "type": "list(string)",
      },
      {
        "default": {},
        "description": "Override default values for target groups. See workers_group_defaults_defaults in local.tf for valid keys.",
        "name": "workers_group_defaults",
        "required": false,
        "type": "any",
      },
      {
        "default": "",
        "description": "User defined workers role name.",
        "name": "workers_role_name",
        "required": false,
        "type": "string",
      },
      {
        "default": true,
        "description": "Whether to write the aws-auth configmap file.",
        "name": "write_aws_auth_config",
        "required": false,
        "type": "bool",
      },
      {
        "default": true,
        "description": "Whether to write a Kubectl config file containing the cluster configuration. Saved to \`config_output_path\`.",
        "name": "write_kubeconfig",
        "required": false,
        "type": "bool",
      },
    ],
    "name": "terraform-aws-modules_eks_aws",
    "outputs": [
      {
        "description": "Name of cloudwatch log group created",
        "name": "cloudwatch_log_group_name",
      },
      {
        "description": "The Amazon Resource Name (ARN) of the cluster.",
        "name": "cluster_arn",
      },
      {
        "description": "Nested attribute containing certificate-authority-data for your cluster. This is the base64 encoded certificate data required to communicate with your cluster.",
        "name": "cluster_certificate_authority_data",
      },
      {
        "description": "The endpoint for your EKS Kubernetes API.",
        "name": "cluster_endpoint",
      },
      {
        "description": "IAM role ARN of the EKS cluster.",
        "name": "cluster_iam_role_arn",
      },
      {
        "description": "IAM role name of the EKS cluster.",
        "name": "cluster_iam_role_name",
      },
      {
        "description": "The name/id of the EKS cluster.",
        "name": "cluster_id",
      },
      {
        "description": "The URL on the EKS cluster OIDC Issuer",
        "name": "cluster_oidc_issuer_url",
      },
      {
        "description": "Security group ID attached to the EKS cluster.",
        "name": "cluster_security_group_id",
      },
      {
        "description": "The Kubernetes server version for the EKS cluster.",
        "name": "cluster_version",
      },
      {
        "description": "A kubernetes configuration to authenticate to this EKS cluster.",
        "name": "config_map_aws_auth",
      },
      {
        "description": "kubectl config file contents for this EKS cluster.",
        "name": "kubeconfig",
      },
      {
        "description": "The filename of the generated kubectl config.",
        "name": "kubeconfig_filename",
      },
      {
        "description": "ARN of the worker autoscaling IAM policy if \`manage_worker_autoscaling_policy = true\`",
        "name": "worker_autoscaling_policy_arn",
      },
      {
        "description": "Name of the worker autoscaling IAM policy if \`manage_worker_autoscaling_policy = true\`",
        "name": "worker_autoscaling_policy_name",
      },
      {
        "description": "default IAM instance profile ARN for EKS worker groups",
        "name": "worker_iam_instance_profile_arns",
      },
      {
        "description": "default IAM instance profile name for EKS worker groups",
        "name": "worker_iam_instance_profile_names",
      },
      {
        "description": "default IAM role ARN for EKS worker groups",
        "name": "worker_iam_role_arn",
      },
      {
        "description": "default IAM role name for EKS worker groups",
        "name": "worker_iam_role_name",
      },
      {
        "description": "Security group ID attached to the EKS workers.",
        "name": "worker_security_group_id",
      },
      {
        "description": "IDs of the autoscaling groups containing workers.",
        "name": "workers_asg_arns",
      },
      {
        "description": "Names of the autoscaling groups containing workers.",
        "name": "workers_asg_names",
      },
      {
        "description": "ID of the default worker group AMI",
        "name": "workers_default_ami_id",
      },
      {
        "description": "ARNs of the worker launch templates.",
        "name": "workers_launch_template_arns",
      },
      {
        "description": "IDs of the worker launch templates.",
        "name": "workers_launch_template_ids",
      },
      {
        "description": "Latest versions of the worker launch templates.",
        "name": "workers_launch_template_latest_versions",
      },
      {
        "description": "User data of worker groups",
        "name": "workers_user_data",
      },
    ],
  },
}
`;

exports[`readSchema generates a single module schema 1`] = `
{
  "terraform-aws-modules_iam_aws_modules_iam-account": {
    "inputs": [
      {
        "description": "AWS IAM account alias for this account",
        "name": "account_alias",
        "required": true,
        "type": "string",
      },
      {
        "default": true,
        "description": "Whether to allow users to change their own password",
        "name": "allow_users_to_change_password",
        "required": false,
        "type": "bool",
      },
      {
        "default": true,
        "description": "Whether to create AWS IAM account password policy",
        "name": "create_account_password_policy",
        "required": false,
        "type": "bool",
      },
      {
        "default": true,
        "description": "Whether to get AWS account ID, User ID, and ARN in which Terraform is authorized",
        "name": "get_caller_identity",
        "required": false,
        "type": "bool",
      },
      {
        "default": false,
        "description": "Whether users are prevented from setting a new password after their password has expired (i.e. require administrator reset)",
        "name": "hard_expiry",
        "required": false,
        "type": "bool",
      },
      {
        "default": 0,
        "description": "The number of days that an user password is valid.",
        "name": "max_password_age",
        "required": false,
        "type": "number",
      },
      {
        "default": 8,
        "description": "Minimum length to require for user passwords",
        "name": "minimum_password_length",
        "required": false,
        "type": "number",
      },
      {
        "default": null,
        "description": "The number of previous passwords that users are prevented from reusing",
        "name": "password_reuse_prevention",
        "required": false,
        "type": "number",
      },
      {
        "default": true,
        "description": "Whether to require lowercase characters for user passwords",
        "name": "require_lowercase_characters",
        "required": false,
        "type": "bool",
      },
      {
        "default": true,
        "description": "Whether to require numbers for user passwords",
        "name": "require_numbers",
        "required": false,
        "type": "bool",
      },
      {
        "default": true,
        "description": "Whether to require symbols for user passwords",
        "name": "require_symbols",
        "required": false,
        "type": "bool",
      },
      {
        "default": true,
        "description": "Whether to require uppercase characters for user passwords",
        "name": "require_uppercase_characters",
        "required": false,
        "type": "bool",
      },
    ],
    "name": "terraform-aws-modules_iam_aws_modules_iam-account",
    "outputs": [
      {
        "description": "The AWS Account ID number of the account that owns or contains the calling entity",
        "name": "this_caller_identity_account_id",
      },
      {
        "description": "The AWS ARN associated with the calling entity",
        "name": "this_caller_identity_arn",
      },
      {
        "description": "The unique identifier of the calling entity",
        "name": "this_caller_identity_user_id",
      },
      {
        "description": "Indicates whether passwords in the account expire. Returns true if max_password_age contains a value greater than 0. Returns false if it is 0 or not present.",
        "name": "this_iam_account_password_policy_expire_passwords",
      },
    ],
  },
}
`;

exports[`readSchema generates a single provider schema 1`] = `
{
  "format_version": "1.0",
  "provider_schemas": {
    "registry.terraform.io/hashicorp/null": {
      "data_source_schemas": {
        "null_data_source": {
          "block": {
            "attributes": {
              "has_computed_default": {
                "computed": true,
                "description": "If set, its literal value will be stored and returned. If not, its value defaults to \`"default"\`. This argument exists primarily for testing and has little practical use.",
                "description_kind": "markdown",
                "optional": true,
                "type": "string",
              },
              "id": {
                "computed": true,
                "deprecated": true,
                "description": "This attribute is only present for some legacy compatibility issues and should not be used. It will be removed in a future version.",
                "description_kind": "markdown",
                "type": "string",
              },
              "inputs": {
                "description": "A map of arbitrary strings that is copied into the \`outputs\` attribute, and accessible directly for interpolation.",
                "description_kind": "markdown",
                "optional": true,
                "type": [
                  "map",
                  "string",
                ],
              },
              "outputs": {
                "computed": true,
                "description": "After the data source is "read", a copy of the \`inputs\` map.",
                "description_kind": "markdown",
                "type": [
                  "map",
                  "string",
                ],
              },
              "random": {
                "computed": true,
                "description": "A random value. This is primarily for testing and has little practical use; prefer the [hashicorp/random provider](https://registry.terraform.io/providers/hashicorp/random) for more practical random number use-cases.",
                "description_kind": "markdown",
                "type": "string",
              },
            },
            "deprecated": true,
            "description": "The \`null_data_source\` data source implements the standard data source lifecycle but does not
interact with any external APIs.

Historically, the \`null_data_source\` was typically used to construct intermediate values to re-use elsewhere in configuration. The
same can now be achieved using [locals](https://www.terraform.io/docs/language/values/locals.html).
",
            "description_kind": "markdown",
          },
          "version": 0,
        },
      },
      "provider": {
        "block": {
          "description_kind": "plain",
        },
        "version": 0,
      },
      "resource_schemas": {
        "null_resource": {
          "block": {
            "attributes": {
              "id": {
                "computed": true,
                "description": "This is set to a random value at create time.",
                "description_kind": "markdown",
                "type": "string",
              },
              "triggers": {
                "description": "A map of arbitrary strings that, when changed, will force the null resource to be replaced, re-running any associated provisioners.",
                "description_kind": "markdown",
                "optional": true,
                "type": [
                  "map",
                  "string",
                ],
              },
            },
            "description": "The \`null_resource\` resource implements the standard resource lifecycle but takes no further action.

The \`triggers\` argument allows specifying an arbitrary set of values that, when changed, will cause the resource to be replaced.",
            "description_kind": "markdown",
          },
          "version": 0,
        },
      },
    },
  },
  "provider_versions": {
    "registry.terraform.io/hashicorp/null": "3.1.0",
  },
}
`;
